### PR TITLE
Fix missing ENOTSUP on PyPy

### LINKED ISSUE #338: Bug: PyPy does not have ENOTSUP
To reproduce: Run the unit tests on PyPy. They fail because `errno` doesn't have ENOTSUP defined, even though it's present on the same platform for CPython.

We can fix it by using the CPython value as a default. Error codes aren't guaranteed to be identical across all platforms so this might not work everywhere, but it'll prevent the crash.

### PR DESCRIPTION
**Pull request by dargueta**
## Type of changes

- [x] Bug fix
- [ ] New feature
- [ ] Documentation / docstrings
- [ ] Tests
- [ ] Other

## Checklist

- [x] I've run the latest [black](https://github.com/ambv/black) with default args on new code.
- [x] I've updated CHANGELOG.md and CONTRIBUTORS.md where appropriate.
- [x] I've added tests for new code.
- [x] I accept that @willmcgugan may be pedantic in the code review.

## Description

1. Only add `errno.ENOTSUP` to FTP failure codes if that's present.
2. Removed a few unused imports.

Closes #338

### CODE REVIEW

**Comment by lurch** [comment]
I wonder if it might be better to fix this with `try ... catch` rather than adding arbitrary error numbers? :man_shrugging:

**Comment by lurch** [comment]
P.S. Is this at all relevant? https://github.com/giampaolo/pysendfile/issues/28

**Comment by dargueta** [comment]
> I wonder if it might be better to fix this with try ... catch rather than adding arbitrary error numbers?

And what would we do on the catch?

> P.S. Is this at all relevant? giampaolo/pysendfile#28

I don't think so?

**Comment by dargueta** [comment]
Crashes in PyPy appear to be unrelated, due to an SFTP connection issue.

File: fs/osfs.py [hunk 1]
```diff
@@ -420,17 +420,18 @@ def _check_copy(self, src_path, dst_path, overwrite=False):
...
+            errno.EINVAL,
+            errno.ENOSYS,
+            errno.EBADF,
+            errno.ENOTSOCK,
+            errno.EOPNOTSUPP,
+        }  # type: MutableSet[int]
+
+        # PyPy doesn't define ENOTSUP so we have to add it conditionally.
+        if hasattr(errno, "ENOTSUP"):
+            _sendfile_error_codes.add(errno.ENOTSUP)
```
Comment (lurch) [inline fs/osfs.py]: I guess you _could_ convert `_sendfile_error_codes` back into a `frozenset` after adding (or not adding) the `errno.ENOTSUP` ?

Comment (dargueta) [reply fs/osfs.py]: Does it need to be one?

Comment (lurch) [reply fs/osfs.py]: Dunno - but I guess it doesn't hurt? I'll leave it up to @willmcgugan to decide :slightly_smiling_face:

File: fs/osfs.py [hunk 2]
```diff
@@ -420,17 +420,18 @@ def _check_copy(self, src_path, dst_path, overwrite=False):
...
-            }
-        )
+        _sendfile_error_codes = {
+            errno.EIO,
+            errno.EINVAL,
+            errno.ENOSYS,
+            errno.EBADF,
+            errno.ENOTSOCK,
+            errno.EOPNOTSUPP,
+        }  # type: MutableSet[int]
```
Comment (willmcgugan) [inline fs/osfs.py]: I think the type is redundant here. Mypy can deduce the types of literals.

File: fs/osfs.py (hunk 1 above)
Comment (willmcgugan) [reply fs/osfs.py]: A regular set will do the job. Don't see any scope for breakage.

File: fs/osfs.py (hunk 2 above)
Comment (lurch) [reply fs/osfs.py]: Does that mean the change on line 63 is also redundant?

Comment (dargueta) [reply fs/osfs.py]: Probably

**Comment by willmcgugan** [comment]
The pypy build is failing on the ftpfs tests. I've seen this before and never figured it out. It's unrelated to your changes.

I'm happy to merge this and treat the pypy test fails as another task. What do you think?

**Comment by dargueta** [comment]
> I'm happy to merge this and treat the pypy test fails as another task.

Conceptually it makes sense, but anyone creating a new PR is gonna be really mad that their PR is broken because of something I did. Up to you.

**Comment by willmcgugan** [comment]
How about remove the pypy builds from travis for now, and I'll add an issue to add them later?

**Comment by willmcgugan** [comment]
Added an issue. https://github.com/PyFilesystem/pyfilesystem2/issues/342

If you wouldn't mind removing the pypy builds I'll merge this...

**Comment by dargueta** [comment]
This is gonna have a merge conflict with the pytest PR #337 which modifies the Travis file and will remove the PyPy tests. Let's get that in first and then this will be good to go.

**Comment by willmcgugan** [comment]
@dargueta Would you mind fixing that merge conflict and we should be good to go.

File: fs/ftpfs.py [hunk 3]
```diff
@@ -728,9 +728,8 @@ def _scandir(self, path, namespaces=None):
                     for raw_info in self._parse_mlsx(lines):
                         yield Info(raw_info)
                     return
-            with self._lock:
-                for info in self._read_dir(_path).values():
-                    yield info
+            for info in self._read_dir(_path).values():
+                yield info
```
Comment (lurch) [inline fs/ftpfs.py]: Does this change have any impact on non-pypy usages of pyfilesystem?

Comment (willmcgugan) [reply fs/ftpfs.py]: I think this was purely because the lock is redundant, as there is a lock on L714.

File: CHANGELOG.md [hunk 4]
```diff
@@ -16,6 +16,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 - Fixed incorrect imports of `mock` on Python 3
 - Removed some unused imports and unused `requirements.txt` file
 - Added mypy checks to Travis
+- Fixed missing `errno.ENOTSUP` on PyPy. Closes [#338](https://github.com/PyFilesystem/pyfilesystem2/issues/338).
```
Comment (lurch) [inline CHANGELOG.md]: Will this PR also close #342 ?

**Comment by willmcgugan** [comment]
Thanks @dargueta !

**Pull request merged**